bennu-io/src/main/java/org/fenixedu/bennu/io/domain/DriveAPIStorage.java

```diff
@@ -0,0 +1,118 @@
+        setName(name);
+        setDriveUrl(driveUrl);
+        setRemoteUsername(remoteUsername);
+        setRemoteDirectoryId(remoteDirectoryId);
+    }
+
+    private transient String accessToken = null;
+    private transient long accessTokenValidUnit = System.currentTimeMillis() - 1;
+
+    private String getAccessToken() {
+        if (accessToken == null || System.currentTimeMillis() >= accessTokenValidUnit) {
+            synchronized (this) {
+                if (accessToken == null || System.currentTimeMillis() >= accessTokenValidUnit) {
+                    final JsonObject claim = new JsonObject();
+                    claim.addProperty("username", getRemoteUsername());
+                    accessToken = Tools.sign(SignatureAlgorithm.RS256, CoreConfiguration.getConfiguration().jwtPrivateKeyPath(), claim);
+                }
+            }
+        }
+        return accessToken;
+    }
+
+    private String uploadFile(final String directory, final Function<MultipartBody, MultipartBody> fileSetter) {
+        final MultipartBody request = Unirest.post(getDriveUrl() + "/api/drive/directory/" + getRemoteDirectoryId())
+                .header("Authorization", "Bearer " + getAccessToken())
+                .header("X-Requested-With", "XMLHttpRequest")
+                .field("path", directory);
+        final HttpResponse<String> response = fileSetter.apply(request).asString();
+        final JsonObject result = new JsonParser().parse(response.getBody()).getAsJsonObject();
+        return result.get("id").getAsString();
+    }
+
+    private InputStream downloadFile(final String fileId) {
+        final HttpResponse<InputStream> response = Unirest.delete(getDriveUrl() + "/api/drive/file/" + fileId + "/download")
+                .header("Authorization", "Bearer " + getAccessToken())
+                .asObject(raw -> raw.getContent());
+        final HttpResponse<InputStream> realResponse = response.getStatus() == 307 ?
+                Unirest.get(response.getHeaders().getFirst("Location")).asObject(raw -> raw.getContent()) : response;
+        return realResponse.getBody();
+    }
+
+    private boolean deleteFile(final String fileId) {
+        final HttpResponse<String> response = Unirest.delete(getDriveUrl() + "/api/drive/file/" + fileId)
+                .header("Authorization", "Bearer " + getAccessToken())
+                .header("X-Requested-With", "XMLHttpRequest")
+                .asString();
+        return response.getStatus() == 204;
+    }
+
+    private static String dirnameFor(final GenericFile file) {
+        final DateTime when = file.getCreationDate();
+        return Integer.toString(when.getYear()) + "/" + when.getMonthOfYear() + "/" + when.getDayOfMonth() + "/"
+                + file.getExternalId();
+    }
+
+    @Override
+    public String store(final GenericFile file, final byte[] content) {
+        if (content == null) {
+            if (!deleteFile(file.getContentKey())) {
```

Comment:
if content is null it behaves as a delete, however the next two store methods don't have the same behavior, why is that?

Reply:
For the same reason it is implemented like this in the LocalFileSystemStorage class.
The store(final GenericFile file, final byte[] content) is the only one called by following method of the GenericFile class:
   `private void setContent(final byte[] content) {`
which in some cases is called by the delete() method. So when the content is a null byte array, we really want to delete the remote file. This is neve the case for the other store methods.